test-programs/compressed-token-test/tests/test.rs

```diff
@@ -432,6 +435,226 @@ async fn test_delegation() {
+            proof: proof_rpc_result.proof.clone(),
+        };
+        let instruction = create_approve_instruction(inputs).unwrap();
+        let context_payer = rpc.get_payer().insecure_clone();
+        rpc.create_and_send_transaction(
+            &[instruction],
+            &sender.pubkey(),
+            &[&context_payer, &sender],
+        )
+        .await
+        .unwrap_err();
+    }
+    // 3. Invalid proof.
+    {
+        let invalid_proof = CompressedProof {
+            a: [0; 32],
+            b: [0; 64],
+            c: [0; 32],
+        };
+
+        let inputs = CreateApproveInstructionInputs {
+            fee_payer: rpc.get_payer().pubkey(),
+            authority: sender.pubkey(),
+            input_merkle_contexts: input_compressed_accounts
+                .iter()
+                .map(|x| x.compressed_account.merkle_context)
+                .collect(),
+            input_token_data: input_compressed_accounts
+                .iter()
+                .map(|x| x.token_data)
+                .collect(),
+            mint,
+            delegated_amount,
+            delegated_compressed_account_merkle_tree,
+            change_compressed_account_merkle_tree: delegated_compressed_account_merkle_tree,
+            delegate: delegate.pubkey(),
+            root_indices: proof_rpc_result.root_indices.clone(),
+            proof: invalid_proof,
+        };
+        let instruction = create_approve_instruction(inputs).unwrap();
+        let context_payer = rpc.get_payer().insecure_clone();
+        let result = rpc
+            .create_and_send_transaction(
+                &[instruction],
+                &sender.pubkey(),
+                &[&context_payer, &sender],
+            )
+            .await;
+        assert_rpc_error(result, 0, VerifierError::ProofVerificationFailed.into()).unwrap();
+    }
+    // 4. Invalid mint.
+    {
+        let invalid_mint = Keypair::new();
+
+        let inputs = CreateApproveInstructionInputs {
+            fee_payer: rpc.get_payer().pubkey(),
+            authority: sender.pubkey(),
+            input_merkle_contexts: input_compressed_accounts
+                .iter()
+                .map(|x| x.compressed_account.merkle_context)
+                .collect(),
+            input_token_data: input_compressed_accounts
+                .iter()
+                .map(|x| x.token_data)
+                .collect(),
+            mint: invalid_mint.pubkey(),
+            delegated_amount,
+            delegated_compressed_account_merkle_tree,
+            change_compressed_account_merkle_tree: delegated_compressed_account_merkle_tree,
+            delegate: delegate.pubkey(),
+            root_indices: proof_rpc_result.root_indices.clone(),
+            proof: proof_rpc_result.proof.clone(),
+        };
+        // NOTE(vadorovsky): Not sure what to do here!
+        // For now, this instruction returns `ProofVerificationFailed`, which
```

Comment:
Right, we talked about it on 1on1. I was previously thinking that it'd be nice to have a dedicated error variant for that, but that'd be an overkill.

Removed the commend and added the assert.

Reply:
Yeah, it's not ideal that we don't have a dedicated error. To throw a dedicated error we would need to send more data unfortunately. We do that in the system program because instruction data is less of a constraint in cpis.